datafusion/core/src/physical_plan/hash_join.rs

```diff
@@ -1054,6 +1110,116 @@ fn equal_rows(
             DataType::LargeUtf8 => {
                 equal_rows_elem!(LargeStringArray, l, r, left, right, null_equals_null)
             }
+            DataType::Decimal(_, lscale) => match r.data_type() {
```

Comment:
From my knowledge in the datafusion, we should convert the left and right to the same data type in the planner.
The data type of left and right must be same.

Reply:
I create two table in the spark，
```
spark-sql> desc t1;
c1                      decimal(10,3)

spark-sql> desc t2;
c1                      decimal(10,4)

and join two table with the eq condition: t1.c1 = t2.c1, get the plan

spark-sql> explain select * from t1,t2 where t1.c1 = t2.c1;
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- SortMergeJoin [cast(c1#111 as decimal(11,4))], [cast(c1#112 as decimal(11,4))], Inner
   :- Sort [cast(c1#111 as decimal(11,4)) ASC NULLS FIRST], false, 0
   :  +- Exchange hashpartitioning(cast(c1#111 as decimal(11,4)), 200), ENSURE_REQUIREMENTS, [id=#293]
   :     +- Filter isnotnull(c1#111)
   :        +- Scan hive default.t1 [c1#111], HiveTableRelation [`default`.`t1`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, Data Cols: [c1#111], Partition Cols: []]
   +- Sort [cast(c1#112 as decimal(11,4)) ASC NULLS FIRST], false, 0
      +- Exchange hashpartitioning(cast(c1#112 as decimal(11,4)), 200), ENSURE_REQUIREMENTS, [id=#294]
         +- Filter isnotnull(c1#112)
            +- Scan hive default.t2 [c1#112], HiveTableRelation [`default`.`t2`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, Data Cols: [c1#112], Partition Cols: []]
```
From the from, we can get the the t1.c1 and t2.c1 will be casted to coerced type(decimal(11,4));

Reply:
I create two tables in datafusion

❯ describe t1;
+-------------+----------------+-------------+
| column_name | data_type      | is_nullable |
+-------------+----------------+-------------+
| c1          | Decimal(10, 2) | NO          |
+-------------+----------------+-------------+

select * from t1;
+------+
| c1   |
+------+
| 1.00 |
+------+


❯ describe t2;
+-------------+----------------+-------------+
| column_name | data_type      | is_nullable |
+-------------+----------------+-------------+
| c1          | Decimal(10, 4) | NO          |
+-------------+----------------+-------------+

select * from t2;
+--------+
| c1     |
+--------+
| 1.0000 |
+--------+

this query gets 0 rows
select * from t1 join t2 on t1.c1 = t2.c1;
0 rows in set. Query took 0.005 seconds.

but this query gets one row
select * from t1 join t2 on t1.c1::decimal(10,4) = t2.c1;
+------+--------+
| c1   | c1     |
+------+--------+
| 1.00 | 1.0000 |
+------+--------+

so I think the planner still not converted the data types between left and right(at least for decimal). perhaps we can add this to planner.
please tell me if I miss something? thanks!

Reply:
We should make the data type is same between left and right.
In this pr, we should better forbid calculate the join operation for diff data type, For example
Decimal(10,3) join Decimal(11,3)
I think you can create a follow up pr to make data type coercion in the planner.

Reply:
> 

yes, we should check the data types and make the coercion in the planner. I'll create an issue and follow up the pr later.

Reply:
Thank you @AssHero